### PR TITLE
Added a custom JVP rule for jax.nn.logsumexp

### LINKED ISSUE #22398: Incorrect Jacobian of `jax.scipy.special.logsumexp` when `b` contains a `0`
### Description

The gradient of `jax.scipy.special.logsumexp` along `b` is wrong when `b` contains a `0`. In the following example, the gradient of $f(x_1, x_2) = \log(x_1 + x_2)$ should be:
```math
\nabla_x f(x) = \left( \frac{1}{x_1 + x_2}, \frac{1}{x_1 + x_2} \right)
```
but JAX outputs that the gradient at `(1,0)` is `(1,0)` when it should be `(1,1)`:

```python
import jax
import jax.numpy as jnp

fun = lambda x: jax.scipy.special.logsumexp(jnp.zeros(2), axis=0, b=x)
a = jnp.array([1, 0], dtype=float)

jax.jacfwd(fun)(a)
```
output:
```python
Array([1., 0.], dtype=float32)
```

There is only a problem for an exact `0`: the gradient taken at `(1,1e-30)` outputs the correct value of `(1,1)`:
```python
a = jnp.array([1, 1e-30], dtype=float)
jax.jacfwd(fun)(a)
```
output:
```python
Array([1., 1.], dtype=float32)
```

Furthermore, if the `logsumexp` is implemented manually, the gradient at `(1, 0)` is also correct:
```python
fun_2 = lambda x: jnp.log(jnp.sum(x * jnp.exp(jnp.zeros(2))))
a = jnp.array([1, 0], dtype=float)
jax.jacfwd(fun_2)(a)
```
output:
```python
Array([1., 1.], dtype=float32)
```

cc @PhilipVinc

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.30
jaxlib: 0.4.30
numpy:  1.24.4
python: 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:37:07) [Clang 15.0.7 ]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Darwin', node='mba-10836208', release='23.4.0', version='Darwin Kernel Version 23.4.0: Fri Mar 15 00:19:22 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T8112', machine='arm64')
```

### PR DESCRIPTION
**Pull request by superbobry**
Fixes #22398 where the Jacobian of jax.nn.logsumexp was wrong if b= contained exact zeros.

### CODE REVIEW

**Comment by superbobry** [comment]
@jakevdp does this look like the fix you had in mind?

File: jax/_src/ops/special.py [hunk 1]
```diff
@@ -95,3 +97,22 @@ def logsumexp(a: ArrayLike, axis: Axis = None, b: ArrayLike | None = None,
...
+  if return_sign:
+    out, sign = out
+
+  if b is None:
+    b = jnp.ones_like(a)
+  out_dot = jnp.exp(
+      logsumexp(a, axis, b * a_dot + b_dot, keepdims, where) - out)
+
+  if return_sign:
+    return (out, sign), out_dot
```
Comment (dfm) [inline jax/_src/ops/special.py:116]: I think this probably needs a `sign_dot` here too. I think the tangents always need to have the same pytree structure as the primals.

File: jax/_src/ops/special.py [hunk 2]
```diff
@@ -95,3 +97,22 @@ def logsumexp(a: ArrayLike, axis: Axis = None, b: ArrayLike | None = None,
...
+def _logsumexp_jvp(axis, keepdims, return_sign, where, primals, tangents):
+  a, b = primals
+  a_dot, b_dot = tangents
+  out = logsumexp(a, axis, b, keepdims, return_sign, where)
+
+  if return_sign:
+    out, sign = out
+
+  if b is None:
+    b = jnp.ones_like(a)
```
Comment (dfm) [inline jax/_src/ops/special.py:111]: We probably need to handle the case where `b_dot` is `None` as well?

File: jax/_src/ops/special.py (hunk 1 above)
Comment (superbobry) [reply jax/_src/ops/special.py:116]: I could be wrong, but I think `sign_dot` is included in `out_dot` if `return_sign` is True.

File: jax/_src/ops/special.py (hunk 2 above)
Comment (superbobry) [reply jax/_src/ops/special.py:111]: Good point. Am I right that `b_dot` is only None if `b` is None?

File: jax/_src/ops/special.py [hunk 3]
```diff
@@ -95,3 +97,22 @@ def logsumexp(a: ArrayLike, axis: Axis = None, b: ArrayLike | None = None,
...
+  out_dot = jnp.exp(
+      logsumexp(a, axis, b * a_dot + b_dot, keepdims, where) - out)
```
Comment (jakevdp) [inline jax/_src/ops/special.py:113]: This expression looks a bit strange to me – each element of `out_dot` should reflect the gradient of the corresponding elements of `a_dot` and `b_dot`, so summing over them doesn't look right to me.

**Comment by jakevdp** [comment]
Let's double-check the formula here. I get this:

$$f(a, b) = \log\sum_i b_i \exp(a_i)$$

$$df_j = da_j \ \partial f/\partial a_j + db_j \ \partial f/\partial b_j$$

$$df_j = da_j b_j\exp(a_j)/\sum_i b_i \exp(a_i) + db_j \exp(a_j)/\sum_i b_i\exp(a_i)$$

$$df_j = (da_j b_j + db_j)\exp(a_j)/\exp(\log\sum_i b_i\exp(a_i))$$

$$df_j = (da_j b_j + db_j)\exp[a_j - f(a, b)]$$

So the out tangent in the general case should be something like
```python
(a_dot * b + b_dot) * jnp.exp(a - out)
```
Does that sound right?

File: jax/_src/ops/special.py [hunk 4]
```diff
@@ -95,3 +97,22 @@ def logsumexp(a: ArrayLike, axis: Axis = None, b: ArrayLike | None = None,
     with jax.debug_nans(False):
       out = jnp.where(sign < 0, jnp.array(np.nan, dtype=out.dtype), out)
   return out
+
+@logsumexp.defjvp
+def _logsumexp_jvp(axis, keepdims, return_sign, where, primals, tangents):
+  a, b = primals
+  a_dot, b_dot = tangents
+  out = logsumexp(a, axis, b, keepdims, return_sign, where)
```
[the remark this answers is not in the document]
Comment (superbobry) [reply jax/_src/ops/special.py:105]: Done.

File: jax/_src/ops/special.py (hunk 2 above)
Comment (superbobry) [reply jax/_src/ops/special.py:111]: Done.

File: jax/_src/ops/special.py (hunk 1 above)
Comment (superbobry) [reply jax/_src/ops/special.py:116]: Added a test with `return_sign=True`.

File: jax/_src/ops/special.py (hunk 3 above)
Comment (superbobry) [reply jax/_src/ops/special.py:113]: I think we do need to sum along `axis`, otherwise the shapes don't add up: `b * a_dot + b_dot` has the same shape as `a`, but the expected `out_dot` shape should be the same as `out`, i.e. lower rank than `a`.

**Comment by superbobry** [comment]
Thanks, Jake! I think this looks right, but it also seems that we need to sum along `axis` to match the `out` shape, which is why I had a `logsumexp`. Wdyt?

**Review by dfm — APPROVED** [review]
Looks good!

**Comment by jakevdp** [comment]
> Thanks, Jake! I think this looks right, but it also seems that we need to sum along `axis` to match the `out` shape, which is why I had a `logsumexp`. Wdyt?

Ah, right. I think `sum` is what we want here. The output tangent should be linear in the input tangents, so taking the exponent and log of the input tangents doesn't sound right.

File: tests/lax_scipy_test.py [hunk 5]
```diff
@@ -202,6 +203,34 @@ def testLogSumExpWhere(self, shape, dtype):
...
+    np.testing.assert_array_equal(
+        jax.jacfwd(fun)(jnp.array([1.0, 0.0])),
+        jnp.ones(2),
+    )
+
+  @parameterized.product(
+      axis=[None, 0], with_b=[False, True], return_sign=[False, True]
+  )
+  def testLogSumExpJac(self, axis, with_b, return_sign):
+    # TODO(slebedev): Test return_sign=True.
```
[the remark this answers is not in the document]
Comment (superbobry) [reply tests/lax_scipy_test.py]: Done.

**Comment by superbobry** [comment]
Discussed with Jake offline:

* We can use `exp(logsumexp(...))`, since exp/log cancel out and the `out_dot` is still linear in the inputs, but
* it is somewhat of a code smell to use non-linear operations on tangents;
* we will get nans if tangents are negative with `logsumexp`.

So, I switched to `sum` and also removed unnecessary allocations for the case of `b is None`.

**Review by jakevdp — APPROVED** [review]
Looks good!
